### ISSUE TITLE
Bootstrap server example requires client TLV support

### ISSUE THREAD

**Issue by qleisan** [eclipse/wakaama#508]
TLV support was made optional in #430 (and the need to build the client with TLV support in order for it to work properly with the bootstrap server example was documented in #502)

The 1.1 spec states (§7.4) that  clients should not use the legacy TLV format. The conclusion is that the bootstrap server needs to be updated to use a different format? 

@sbertin-telular do you have input on this?

**Comment by sbertin-telular** [comment]
I think this is only a temporary requirement until the changes to support LWM2M 1.1 for bootstrap are in place. I think the changes I have on https://github.com/sbertin-telular/wakaama/tree/bootstrap_format will eliminate the requirement once merged.

**Comment by sbernard31** [comment]
I suppose a bootstrap server should be able to handle LWM2M 1.0 and LWM2M 1.1 client.
So content format to use SHOULD probably depends on version used by the client.
AFAIK, the client does not send the LWM2M version on `Bootstrap Request` but in LWM2M 1.1 there is a new bootstrap parameter : `Preferred Content Format`.

So when this parameter is present that means that client is LWM2M 1.1 and by the way we know which content format to use.

The issue is this parameter is NOT required... So when client don't use it, we don't know if we face a LWM2M 1.0 or LWM2M 1.1 client...I suppose in this case the probability that we face either a "LWM2M 1.0" client or a "LWM2M 1.1 client which support TLV" is higher ans so we should use TLV ?
Or we need to begin bootstrap by a `Discover Request` ?

**Comment by qleisan** [comment]
In our client (default 1.1) we should definitely supply the the parameter so client and bootstrap_server can communicate without the need to build the client with TLV support. I suspect this is handled by sbertin-telulars branch.

@sbertin-telular does your branch include full support for the "format option" on the receiving end and include it on the sending end? If the bootstrap server still use TLV when the option is not supplied (e.g. a 1.0 client) it should be fine.
In summary - do you see any case not handled by your work?

**Comment by sbertin-telular** [comment]
The specific commit is https://github.com/sbertin-telular/wakaama/commit/04e821972db2f5146b267f62cb8203bcc5a40156. For LWM2M 1.1 the client will always provide the ``Preferred Content Format``. The server will default to TLV if it does not support the ``Preferred Content Format`` from the client. An LWM2M 1.1 client that does not send the ``Preferred Content Format`` and does not support TLV would be a problem. Working around it would require processing the error response and trying a different format until one works. That would require quite a bit of additional logic.

**Comment by sbernard31** [comment]
I think this is a good behavior. :+1:

About,
> An LWM2M 1.1 client that does not send the Preferred Content Format and does not support TLV would be a problem.

I think a client which does not support TLV should seriously consider to use `prefered content format`. (I don't know why this is not mandatory?)
My Idea about using Discover Request does not help so much because there is no mandatory "multi value" Content format in LWM2M 1.1 ... (I just discover this reading the spec again)

```
The LwM2M Client MUST support plain text, opaque, and CoRE Link data formats and it SHOULD support at least one of SenML CBOR or SenML JSON data formats. 
```

So your idea about processing error response and trying a different format is probably the only way ? :thinking: 

In LWM2M 1.0 the server and the client MUST support TLV and TLV can be used for almost all needs. (single value, opaque, multi value) it seems to me this was really better for interoperability ? Now in LWM2M 1.1 it's more difficult to know which content format must be used by default ? I'm talking about bootstrap but also LWM2M server, or maybe I missed something ?

**Comment by qleisan** [comment]
In my opinion there is no need to introduce that additional logic now (or possibly never). Great that there is already a solution waiting for this bootstrap TLV issue.

**Comment by sbernard31** [comment]
Should we close this or ?

**Comment by sbertin-telular** [comment]
I think leave it open until the code is merged.

**Comment by sbernard31** [comment]
(An OMA issue which is related to this : https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues/515)

**Comment by sbertin-telular** [comment]
I think we can close this now. The code was merged by #539.
